lib/internal/inspector/_inspect.js

```diff
@@ -44,88 +77,72 @@ class StartupError extends Error {
-    }, timeout);
-
-    function pingPort() {
-      if (didTimeOut) return;
+  setTimeout(timeout).then(() => ac.abort());
 
+  // eslint-disable-next-line no-unused-vars
+  for await (const _ of setInterval(retryDelay)) {
+    if (signal.aborted) {
+      throw new StartupError(
+        `Timeout (${timeout}) waiting for ${host}:${port} to be free`);
+    }
+    const error = await new Promise((resolve) => {
       const socket = net.connect(port, host);
-      let didRetry = false;
-      function retry() {
-        if (!didRetry && !didTimeOut) {
-          didRetry = true;
-          setTimeout(pingPort, retryDelay);
-        }
-      }
-
-      socket.on('error', (error) => {
-        if (error.code === 'ECONNREFUSED') {
-          resolve();
-        } else {
-          retry();
-        }
-      });
-      socket.on('connect', () => {
-        socket.destroy();
-        retry();
-      });
+      socket.on('error', resolve);
+      socket.on('connect', resolve);
+    });
+    if (error?.code === 'ECONNREFUSED') {
+      return;
     }
-    pingPort();
-  });
+  }
 }
 
-function runScript(script, scriptArgs, inspectHost, inspectPort, childPrint) {
-  return portIsFree(inspectHost, inspectPort)
-    .then(() => {
-      return new Promise((resolve) => {
-        const needDebugBrk = process.version.match(/^v(6|7)\./);
-        const args = (needDebugBrk ?
-          ['--inspect', `--debug-brk=${inspectPort}`] :
-          [`--inspect-brk=${inspectPort}`])
-          .concat([script], scriptArgs);
-        const child = spawn(process.execPath, args);
-        child.stdout.setEncoding('utf8');
-        child.stderr.setEncoding('utf8');
-        child.stdout.on('data', (chunk) => childPrint(chunk, 'stdout'));
-        child.stderr.on('data', (chunk) => childPrint(chunk, 'stderr'));
-
-        let output = '';
-        function waitForListenHint(text) {
-          output += text;
-          if (/Debugger listening on ws:\/\/\[?(.+?)\]?:(\d+)\//.test(output)) {
-            const host = RegExp.$1;
-            const port = Number.parseInt(RegExp.$2);
-            child.stderr.removeListener('data', waitForListenHint);
-            resolve([child, port, host]);
-          }
-        }
-
-        child.stderr.on('data', waitForListenHint);
-      });
-    });
+const debugRegex = /Debugger listening on ws:\/\/\[?(.+?)\]?:(\d+)\//;
+async function runScript(script, scriptArgs, inspectHost, inspectPort,
+                         childPrint) {
+  await portIsFree(inspectHost, inspectPort);
+  const args = ArrayPrototypeConcat(
+    [`--inspect-brk=${inspectPort}`, script],
+    scriptArgs);
+  const child = spawn(process.execPath, args);
+  child.stdout.setEncoding('utf8');
+  child.stderr.setEncoding('utf8');
+  child.stdout.on('data', (chunk) => childPrint(chunk, 'stdout'));
+  child.stderr.on('data', (chunk) => childPrint(chunk, 'stderr'));
+
+  let output = '';
+  return new Promise((resolve) => {
```

Comment:
Should this go ahead and use `return await Promise` so that any possible error stack would include the `runScript` frame?

Reply:
I'm not sure I follow, isn't `await <promise>` and `return <promise>` similar when it's the last expression on an `async` function?

Reply:
Yes and no. With `return <promise>` if `<promise>` rejects, the error stack trace will not include the wrapping async function (in this case `runScript`). However, when using `return await <promise>` the error stack *will* contain `runScript`.

Reply:
Am I doing this wrong? It seems the stack trace does contain the wrapping function in the dummy test cases I've came up with..
![V8 DevTools showing stack traces](https://user-images.githubusercontent.com/14309773/116319184-59f44800-a7b6-11eb-903d-5ef3780deefe.png)

Reply:
Yeah looking at it again, in this particular case returning the Promise is ok. Here's the case that I was talking about:

```js
const { setTimeout } = require('timers/promises');

async function foo() {
  await setTimeout(100);
  return Promise.reject(new Error('boom'));
}

async function bar() {
  return foo();
}

// vs.

async function bar2() {
  return await foo();
}
```

Running `bar()` produces an error:

```
> Uncaught Error: boom
    at foo (REPL33:4:23)
```

While running `bar2()` produces an error:

```
> Uncaught Error: boom
    at foo (REPL33:4:23)
    at async bar2 (REPL41:2:8)
```

Reply:
Seems like a V8 bug, interesting. Unless you feel strongly about this, I'm tempted to leave it as is until a linter rule comes to enforce it.